src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java

```diff
@@ -42,6 +45,15 @@ public abstract class AbstractPathTestSupport {
      */
     protected abstract String getPackageLocation();
 
+    /**
+     * Sets the English locale for all tests.
+     * Otherwise, some tests failed in other locales.
+     */
+    @BeforeAll
+    public static void setEnglishLocale() {
+        Locale.setDefault(Locale.ENGLISH);
```

Comment:
This is turning into an ugly issue. `ViolationTest` has a static final Locale which gets its value before this method is run. So now we can have a test global variable with its own Locale and then this Before method comes in and set it to its own specific value and possibly conflict with each other. If they both have the same value, there is no issue, but if they are different then this could become messy.

@romani @nick-mancuso In my own project, I had to create a rule that "static final fields" for tests must be a single state throughout the entire life of the tests and can never change. I really found this to be helpful when I introduced random test ordering (https://github.com/checkstyle/checkstyle/issues/4818) as I found bad tests that were modifying values that should have been constant throughout the entire test run and was causing unpredictable issues.  I also forbid mutable fields for tests so 1 test can be completely isolated, but I am willing for this to be ignored for now.

I think we need to incorporate this rule and reconsider these fields, probably starting with this PR.

Reply:
https://github.com/checkstyle/checkstyle/blob/fe3969a2f16854a93d66e80e0ac2df3aa5a3102c/src/test/java/com/puppycrawl/tools/checkstyle/api/ViolationTest.java#L55

Is used only 3 times , we can move definition to each test , to let test control locale .

Goal of this PR to let people with non EN locale to run tests/build without problems. Benefits of this is singnificatly bigger than potential problem that we can resolve later on by overrides and any other way.

Reply:
All tests which want to play with locale can have separate base class , and do what makes sense to them.

Reply:
> we can move definition to each test 

It will most likely be more involved than this. We are setting the default Locale probably after all our production classes have loaded. Some like Violation, use the preset locale which is loaded before the before test is run. This is starting to look like all tests need to extend a base and we need a basic init routine.

Reply:
My suggetion add `@DefaultLocale("en")` in each test

This is convenient for those who have a different locale, if you need to locally check that the message is correct in your language, just change locale in test

Reply:
This won't properly handle our production code which is what I was referring to in my previous post.
Example: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java#L68

Just for noting, `@DefaultLocale` saves the current default locale before the test is run, switches the locale to the specified in the annotation, and the restores the previous default locale after the test is run.

Reply:
IMO, I don't think we should do this BeforeEach and setting the locale in it. We will have to setup and maintain more than just the locale for our entire suite to be in that language.

The whole reason for setting the locale to english is because some tests require english language to validate properly.

My suggestion is we modify tests to skip validation when English is required and language in use is not English. The environment is already in place that the language coming in is used everywhere. We fixed the issue with our CI to run tests in any locale. Lets just skip tests that test English specific results.

Reply:
Another reason to not use `Locale.setDefault`. https://github.com/checkstyle/checkstyle/issues/7436

Reply:
> Lets just skip tests that test English specific results.

as we use english messages in Input files, we are forced to run in EN locale.

Reply:
@rnveach 

>Before method comes in and set it to its own specific value and possibly conflict with each other. If they both have the same value, there is no issue, but if they are different then this could become messy.

Are we able to make assertions about this case (different locale)?

>My suggestion is we modify tests to skip validation when English is required and language in use is not English.

can you expand on this?

Reply:
> Are we able to make assertions about this case (different locale)?

If you are talking about code assertions, then yea they can always be added, but it doesn't prevent the issue. And then another issue is what if new code comes in. We won't remember to add more of these checks and the issue could become worse, that is why this isn't a good change.

> My suggestion is we modify tests to skip validation when English is required and language in use is not English.
> can you expand on this?

From what I remember of the issue, basically when a non-English user runs our tests there are failures related to language. Some test(s) is expecting English, and the result is actually the user's non-English locale.

I believe my suggestion was to drop all the code in this PR, and skip over the current test failures ONLY if the user's locale isn't English and validate as many other things as we can. This wouldn't skip anything in CI, as CI doesn't fail on this issue currently.
Basically:
````
if (language == English) {
  validate(....);
}
````

My other suggestion is to actually fix the issue, which is what my other PRs (https://github.com/checkstyle/checkstyle/pull/12511 and https://github.com/checkstyle/checkstyle/pull/12560) were doing. In those PRs, since Checker controls the locale currently (https://checkstyle.org/config.html#Checker_Properties), it would pass it to all the other modules for when a locale is needed and we would drop the global static which is the center of the issue. Basically switching a global static, to a stack based.